scripts/exporter.py

```diff
@@ -85,25 +159,368 @@ def main(self) -> None:
+        )
+
+        # create the point cloud
+        point_cloud = []
+        for i, _ in enumerate(depth_images):
+            ray_bundle = cameras.generate_rays(camera_indices=i)
+            points = ray_bundle.origins + ray_bundle.directions * depth_images[i]
+            point_cloud.append(points)
+        point_cloud = torch.stack(point_cloud)
+        point_cloud = point_cloud.reshape(-1, 3)
+        color_cloud = torch.from_numpy(np.stack(color_images)).reshape(-1, 3)
+
+        # TODO: clip the point cloud based on the AABB
+        # currently this clips between -1 and 1
+        valid = (
+            (point_cloud[:, 0] > -1.0)
+            & (point_cloud[:, 0] < 1.0)
+            & (point_cloud[:, 1] > -1.0)
+            & (point_cloud[:, 1] < 1.0)
+            & (point_cloud[:, 2] > -1.0)
+            & (point_cloud[:, 2] < 1.0)
+        )
+        point_cloud = point_cloud[valid]
+
+        pcd = o3d.geometry.PointCloud()
+        pcd.points = o3d.utility.Vector3dVector(point_cloud.float().numpy())
+        pcd.colors = o3d.utility.Vector3dVector(color_cloud[valid].float().numpy())
+
+        # with CONSOLE.status("Cleaning Point Cloud", spinner="circleHalves") as status:
+        CONSOLE.print("Cleaning Point Cloud")
+        pcd, _ = pcd.remove_statistical_outlier(nb_neighbors=20, std_ratio=1.0)
+        print("\033[A\033[A")
+        CONSOLE.print("[bold green]:white_check_mark: Cleaning Point Cloud")
+
+        # with CONSOLE.status("Estimating Normals", spinner="arrow") as status:
+        CONSOLE.print("Estimating Normals")
+        pcd.estimate_normals()
+        print("\033[A\033[A")
+        CONSOLE.print("[bold green]:white_check_mark: Estimating Normals")
+
+        # with CONSOLE.status("Cleaning Normals", spinner="arrow"):
+        #     pcd.orient_normals_consistent_tangent_plane(100)
+
+        # with CONSOLE.status("Saving Point Cloud", spinner="pipe") as status:
+        CONSOLE.print("Saving Point Cloud")
+        o3d.io.write_point_cloud(str(output_dir / "point_cloud.ply"), pcd)
+        print("\033[A\033[A")
+        CONSOLE.print("[bold green]:white_check_mark: Saving Point Cloud")
+
+        # with CONSOLE.status("Computing Mesh", spinner="aesthetic") as status:
+        CONSOLE.print("Computing Mesh")
+        mesh, densities = o3d.geometry.TriangleMesh.create_from_point_cloud_poisson(pcd, depth=9)
+        vertices_to_remove = densities < np.quantile(densities, 0.1)
+        mesh.remove_vertices_by_mask(vertices_to_remove)
+        print("\033[A\033[A")
+        CONSOLE.print("[bold green]:white_check_mark: Computing Mesh")
+
+        # with CONSOLE.status("Saving Mesh", spinner="pipe") as status:
+        CONSOLE.print("Saving Mesh")
+        o3d.io.write_triangle_mesh(str(output_dir / "mesh.ply"), mesh)
+        print("\033[A\033[A")
+        CONSOLE.print("[bold green]:white_check_mark: Saving Mesh")
+
+
+@dataclass
+class ExportTextureMesh(Exporter):
```

Comment:
Just fyi if you want to do proper texturing I'd suggest using https://github.com/nmoehrle/mvs-texturing -- there are no python bindings so you'd have to ask people to build from source and then point to the build folder, but will probably give you the best results and pretty fast

Reply:
Be aware that MVS Texturing sometimes has strange color shifts.
Meshroom can also do texturing, but the the texturing part is a bit harder to run.

Reply:
Ok, fwiw, I've never had any issues with color shifts and I've used this a lot. The texturing quality on openMVS is slightly higher: https://github.com/cdcseacave/openMVS, although there are more third party dependencies so if you only need texturing then mvs-texturing is an easier integration. They are all based on this paper: https://link.springer.com/content/pdf/10.1007/978-3-319-10602-1_54.pdf

Reply:
Thanks @cpheinrich for the pointers!

Reply:
Maybe you're right @cpheinrich, this issue was solved since https://github.com/nmoehrle/mvs-texturing/issues/24
However, this is still not optimizing the reprojected images as NeRF does (and tonemapping mode makes it a bit more unreal).

What would better fit in nerfstudio is something where we would get the texture from NeRF (even though it's lower resolution):
1. produce the mesh using some method (can be density or pointcloud or something else)
2. create UVs for the mesh (eg using xatlas)
3. render a geometry image (a render of XYZ positions in UV space) -> this gives you XYZ for each texture point
4. compute texture by evaluating nerf at each 3D point, maybe averaging over directions that are on the outer side of the surface

Before step 4 we could fine-tune the NeRF, as done in [Surface NeRF aka SNeRF](https://arxiv.org/abs/2108.04886) - this is called SNeRF (fixed) in the paper.

Step 4 could be replaced by something where we compute another representation for a surface lightfield, eg spherical harmonics,

Reply:
@cpheinrich @devernay

I updated the description of the PR with what we are doing for the first pass. Our UV texture image is a "texture atlas" since every face maps to a unique location. We then color the texture image with NeRF.

Averaging over directions (step 4. suggested by @devernay) will likely give better results.

Happy to hear thoughts on this approach and what can be improved. Thanks!

This is the result I'm getting with TSDF + this type of texturing.
https://user-images.githubusercontent.com/16965789/202813416-9d6eeed7-6e7f-4f0d-b2ad-356a460c1787.mov